ocaml/xapi/repository.ml

```diff
@@ -769,67 +792,157 @@ let set_pending_guidances ~__context ~host ~guidances =
              (Guidance.to_string g)
        )
 
+let apply_livepatches' ~__context ~host ~livepatches =
+  List.filter_map
+    (fun (lp, _) ->
+      let component_str =
+        Livepatch.string_of_component lp.LivePatch.component
+      in
+      try
+        Helpers.call_api_functions ~__context (fun rpc session_id ->
+            let open LivePatch in
+            Client.Client.Repository.apply_livepatch ~rpc ~session_id ~host
+              ~component:component_str ~base_build_id:lp.base_build_id
+              ~base_version:lp.base_version ~base_release:lp.base_release
+              ~to_version:lp.to_version ~to_release:lp.to_release
+        ) ;
+        None
+      with
+      | Api_errors.(Server_error (_, _)) as e ->
+          let host' = Ref.string_of host in
+          let msg = ExnHelper.string_of_exn e in
+          error "applying %s livepatch (%s) on host ref='%s failed: %s"
+            component_str (LivePatch.to_string lp) host' msg ;
+          Some (lp, e)
+      | e ->
+          let host' = Ref.string_of host in
+          let msg = ExnHelper.string_of_exn e in
+          error "applying %s livepatch (%s) on host ref='%s failed: %s"
+            component_str (LivePatch.to_string lp) host' msg ;
+          Some (lp, Api_errors.(Server_error (apply_livepatch_failed, [msg])))
+    )
+    livepatches
+
+let apply_updates' ~__context ~host ~updates_info ~livepatches ~acc_rpm_updates
+    =
+  (* This function runs on coordinator host *)
+  (* Install RPM updates firstly *)
+  Helpers.call_api_functions ~__context (fun rpc session_id ->
+      Client.Client.Repository.apply ~rpc ~session_id ~host
+  ) ;
+  (* Apply live patches with best effort *)
+  let failed_livepatches = apply_livepatches' ~__context ~host ~livepatches in
```

Comment:
Using [Either](https://v2.ocaml.org/api/Either.html) for `apply_livepatches'` would allow using [List.partition_map](https://v2.ocaml.org/api/List.html#VALpartition_map), enhancing clarity and efficiency:

```
let apply_livepatches ... =
  List.partition_map
    (fun (lp, _) ->
      try ...
        Left lp
      with
      | ... ->
          Right (lp, e)
      | ... -> ...
    )
    livepatches
[...]

    let successful_livepatches, failed_livepatches = apply_livepatches' ~__context ... in
    ...